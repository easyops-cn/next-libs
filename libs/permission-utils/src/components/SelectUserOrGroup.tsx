import React from "react";
import { Select } from "antd";
import { filter, isEmpty } from "lodash";
import {
  SelectUserOrGroupProps,
  SelectUserOrGroupState,
  User,
} from "../interfaces";
import { InstanceApi_postSearch } from "@next-sdk/cmdb-sdk";
import { handleHttpError } from "@next-core/brick-kit";
import { LabeledValue } from "antd/lib/select";

export class SelectUserOrGroup extends React.Component<
  SelectUserOrGroupProps,
  SelectUserOrGroupState
> {
  constructor(props: SelectUserOrGroupProps) {
    super(props);
    this.state = {
      users: [],
      userGroups: [],
    };
  }
  fetchUser() {
    return InstanceApi_postSearch("USER", {
      query: {},
      fields: {
        instanceId: true,
        name: true,
      },
    });
  }
  fetchUserGroup() {
    return InstanceApi_postSearch("USER_GROUP", {
      query: {},
      fields: {
        instanceId: true,
        name: true,
      },
    });
  }
  async componentDidMount() {
    try {
      const [users, groups] = await Promise.all([
        this.fetchUser(),
        this.fetchUserGroup(),
      ]);
      this.setState({
        users: users.list as User[],
        userGroups: groups.list as User[],
      });
    } catch (e) {
      handleHttpError(e);
    }
  }
  handleUsersChange = (value: LabeledValue[]) => {
    this.props.handleUsersChange(value);
  };
  filterOpts = (currentUsers: string[], allUsers: User[]) => {
    return filter(allUsers, (item) => currentUsers.includes(item.name));
  };
  render(): React.ReactNode {
    const { currentUsers } = this.props;
    const { users: allUsers, userGroups: allUserGroups } = this.state;
    const users = isEmpty(currentUsers)
      ? allUsers
      : this.filterOpts(currentUsers, allUsers);
    const userGroups = isEmpty(currentUsers)
      ? allUserGroups
      : this.filterOpts(currentUsers, allUserGroups);
    return (
      <Select
        labelInValue
        style={{ width: "100%" }}
        mode="multiple"
        placeholder="选择用户（组）"
        onChange={this.handleUsersChange}
        optionFilterProp="label"
      >
        <Select.OptGroup label="用户">
          {users.map((item: any) => (
            <Select.Option
              value={item.instanceId}
              key={item.instanceId}
              label={item.name}
            >
              {item.name}
            </Select.Option>
          ))}
        </Select.OptGroup>
        <Select.OptGroup label="用户组">
          {userGroups.map((item: any) => (
            <Select.Option
              value={item.instanceId}
              key={item.instanceId}
              label={item.name}
            >
              {item.name}
            </Select.Option>
          ))}
        </Select.OptGroup>
      </Select>
    );
  }
}
