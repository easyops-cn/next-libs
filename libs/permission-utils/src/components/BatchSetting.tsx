import React from "react";
import { QuestionCircleOutlined } from "@ant-design/icons";
import {
  Button,
  Modal,
  Checkbox,
  Switch,
  Alert,
  Radio,
  Row,
  Col,
  Tooltip,
} from "antd";
import { SelectUserOrGroup } from "./SelectUserOrGroup";
import { initPermissionOptions } from "../processors";
import { KEY_AUTHORIZERS_OF_PERM } from "../constants";
import styles from "./index.module.css";
export interface BatchSettingProps {
  modelData?: any;
  instanceIds: string[];
  permissionList: any;
  updateFunction(data: {
    authorizers: string[];
    method: string;
    permissions: [];
    packageIds: string[];
  }): void;
}
export interface BatchSettingState {
  visible: boolean;
  formData: {
    authorizers: string[];
    perm: [];
    enableWhiteList: boolean;
    method: string;
  };
}
export class BatchSetting extends React.Component<
  BatchSettingProps,
  BatchSettingState
> {
  constructor(props: BatchSettingProps) {
    super(props);
    this.state = {
      visible: false,
      formData: {
        enableWhiteList: false,
        method: "overwrite",
        authorizers: [],
        perm: [],
      },
    };
  }
  // 弹出
  showModal = () => {
    this.setState({ visible: true });
  };
  handleOk = () => {
    const packageIds = this.props.instanceIds;
    const { authorizers, method, perm: permissions } = this.state.formData;
    this.props.updateFunction({ authorizers, method, permissions, packageIds });
    this.handleCancel();
  };
  handleCancel = () => {
    this.setState((prevState: BatchSettingState) => ({
      visible: false,
      formData: {
        ...prevState.formData,
        enableWhiteList: false,
      },
    }));
  };
  handleSwitchWhiteList = (checked: boolean) => {
    this.setState((prevState: BatchSettingState) => ({
      formData: {
        ...prevState.formData,
        enableWhiteList: true,
      },
    }));
  };
  handleUsersChange = (value: { key: string; label: string }[]) => {
    const authorizers = value.map(
      (item: { key: string; label: string }) => item.label
    );
    this.setState((prevState: BatchSettingState) => ({
      formData: { ...prevState.formData, authorizers },
    }));
  };
  handleChangeAction = (e: any) => {
    this.setState((prevState: BatchSettingState) => ({
      formData: { ...prevState.formData, method: e.target.value },
    }));
  };
  handleCheckPerm = (checkedValue: any) => {
    this.setState((prevState: BatchSettingState) => ({
      formData: { ...prevState.formData, perm: checkedValue },
    }));
  };
  renderWhiteListForm() {
    const RadioGroup = Radio.Group;
    const RadioButton = Radio.Button;
    const method = this.state.formData.method;
    const operations = [
      {
        label: "重置",
        value: "overwrite",
        message: "重置白名单为：",
      },
      {
        label: "添加",
        value: "append",
        message: "向白名单添加：",
      },
      {
        label: "移除",
        value: "remove",
        message: "从白名单移除：",
      },
    ];
    const message = operations.find((item) => item.value === method).message;

    return (
      <div className={styles.whiteListFormContainer}>
        <div>
          <RadioGroup
            onChange={this.handleChangeAction}
            defaultValue="overwrite"
          >
            {operations.map((operation) => (
              <RadioButton value={operation.value} key={operation.value}>
                {operation.label}
              </RadioButton>
            ))}
          </RadioGroup>
        </div>
        <div className={styles.subFormRow}>
          <label className={styles.subFormLabel}>{message}</label>
          <div>
            <SelectUserOrGroup handleUsersChange={this.handleUsersChange} />
          </div>
        </div>
      </div>
    );
  }
  formIsValid = () => {
    const { perm, enableWhiteList, authorizers } = this.state.formData;
    return perm.length && (!enableWhiteList || authorizers.length);
  };
  render(): React.ReactNode {
    let { modelData, instanceIds } = this.props;
    if (!instanceIds) {
      instanceIds = [];
    }
    const permissionOptions = initPermissionOptions(this.props.permissionList);
    const CheckboxGroup = Checkbox.Group;
    const whiteListForm = this.renderWhiteListForm();
    return (
      <div>
        <Button onClick={this.showModal} disabled={!instanceIds.length}>
          批量设置权限
        </Button>
        <Modal
          title="批量设置权限"
          visible={this.state.visible}
          onOk={this.handleOk}
          onCancel={this.handleCancel}
          okText="确定"
          cancelText="取消"
          okButtonProps={{ disabled: !this.formIsValid() }}
          destroyOnClose={true}
        >
          <Alert
            message={`已选择${instanceIds.length}个${
              modelData ? modelData.name : "程序包"
            }`}
            type="info"
          />
          <div className={styles.formContainer}>
            <div className={styles.formRow}>
              <label
                className={`ant-legacy-form-item-required ${styles.formLabel}`}
              >
                选择权限：
              </label>
              <CheckboxGroup
                options={permissionOptions}
                onChange={this.handleCheckPerm}
              />
            </div>
            <div className={styles.formRow}>
              <label className={styles.formLabel}>白名单：</label>
              <div>
                <Switch
                  defaultChecked={false}
                  onChange={this.handleSwitchWhiteList}
                />
                {this.state.formData.enableWhiteList ? whiteListForm : null}
              </div>
            </div>
          </div>
        </Modal>
      </div>
    );
  }
}
