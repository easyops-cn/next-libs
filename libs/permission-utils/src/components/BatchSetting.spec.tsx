import React from "react";
import { shallow, mount } from "enzyme";
import { BatchSetting } from "./BatchSetting";
import { Button, Modal, Checkbox, Switch, Alert, Radio } from "antd";
import { SelectUserOrGroup } from "./SelectUserOrGroup";
describe("BatchSetting", () => {
  const modelData = { objectId: "APP", name: "应用" };
  const permissionList = [
    {
      action: "deploy:package_delete",
      id: "5b7b87a27e5a2e22bb29bcbf",
      remark: "包删除",
      resource: {
        system: "deploy",
        condition: {
          deleteAuthorizers: "%user",
          name: "package",
          system: "deploy",
        },
      },
      roles: [
        "系统管理员",
        "应用运维",
        "应用测试",
        "应用开发",
        "test",
        "测试角色-没有easyops+系统管理菜单",
      ],
      system: "持续部署",
      user: [
        "willniu123",
        "easyops",
        "qw",
        "qwe",
        "ea",
        "willniu",
        "lightjiao",
        "user3",
      ],
      _actionWeight: 1,
      _category: "a-0",
    },
    {
      action: "deploy:package_update",
      id: "5b7b87a27e5a2e22bb29bcc0",
      remark: "包编辑",
      resource: {
        system: "deploy",
        condition: {
          updateAuthorizers: "%user",
          name: "package",
          system: "deploy",
        },
      },
      roles: [
        "系统管理员",
        "应用运维",
        "应用测试",
        "应用开发",
        "test",
        "测试角色-没有easyops+系统管理菜单",
      ],
      system: "持续部署",
      user: [
        "willniu123",
        "easyops",
        "qw",
        "qwe",
        "ea",
        "willniu",
        "lightjiao",
        "user3",
      ],
      _actionWeight: 2,
      _category: "a-0",
    },
    {
      action: "deploy:package_read",
      id: "5b7b87a27e5a2e22ba29bcc4",
      remark: "包查看",
      resource: {
        system: "deploy",
        condition: {
          readAuthorizers: "%user",
          name: "package",
          system: "deploy",
        },
      },
      roles: [
        "系统管理员",
        "应用运维",
        "应用测试",
        "应用开发",
        "test",
        "测试角色-没有easyops+系统管理菜单",
      ],
      system: "持续部署",
      user: [
        "willniu123",
        "easyops",
        "qw",
        "qwe",
        "ea",
        "willniu",
        "lightjiao",
        "user3",
      ],
      _actionWeight: 100,
      _category: "a-0",
    },
  ];
  const instanceIds = [
    "b6c5145b627a11e99f5106e69ba71c45",
    "8d0eb0decf6a34debdb3c5d458df5f98",
    "5a8c4609657c11e985d58e589cd9215b",
  ];
  const updateFunction = (): void => void 0;
  const props = {
    modelData,
    instanceIds,
    permissionList,
    updateFunction,
  };
  const wrapper = shallow(<BatchSetting {...props} />);
  const instance = wrapper.instance() as BatchSetting;

  it("should work", () => {
    expect(wrapper).toBeTruthy();
    expect(instance.state).toEqual({
      visible: false,
      formData: {
        enableWhiteList: false,
        method: "overwrite",
        authorizers: [],
        perm: [],
      },
    });
  });
  it("should work when toggle the white list switch", () => {
    wrapper.find(Switch).prop("onChange");
    expect(instance.state.formData.enableWhiteList).toEqual(false);
  });
  it("should set authorizers", () => {
    instance.handleUsersChange([{ label: "abc", key: "12345" }]);
    expect(instance.state.formData.authorizers).toEqual(["abc"]);
  });
  it("should set action", () => {
    instance.handleChangeAction({ target: { value: "overwrite" } });
    expect(instance.state.formData.method).toEqual("overwrite");
  });
  it("should set perms", () => {
    instance.handleCheckPerm(["deploy:package_delete"]);
    expect(instance.state.formData.perm).toEqual(["deploy:package_delete"]);
  });
});
