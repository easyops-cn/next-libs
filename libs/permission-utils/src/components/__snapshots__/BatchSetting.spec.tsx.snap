// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchSetting should work 1`] = `
<div>
  <Button
    block={false}
    disabled={false}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[Function]}
  >
    批量设置权限
  </Button>
  <Modal
    cancelText="取消"
    confirmLoading={false}
    destroyOnClose={true}
    maskTransitionName="fade"
    okButtonProps={
      Object {
        "disabled": true,
      }
    }
    okText="确定"
    okType="primary"
    onCancel={[Function]}
    onOk={[Function]}
    title="批量设置权限"
    transitionName="zoom"
    visible={false}
    width={520}
  >
    <Alert
      message="已选择3个应用"
      type="info"
    />
    <div
      className="formContainer"
    >
      <div
        className="formRow"
      >
        <label
          className="ant-form-item-required formLabel"
        >
          选择权限：
        </label>
        <CheckboxGroup
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "包删除",
                "value": "deploy:package_delete",
              },
              Object {
                "label": "包编辑",
                "value": "deploy:package_update",
              },
              Object {
                "label": "包查看",
                "value": "deploy:package_read",
              },
            ]
          }
        />
      </div>
      <div
        className="formRow"
      >
        <label
          className="formLabel"
        >
          白名单：
          <Tooltip
            arrowPointAtCenter={false}
            autoAdjustOverflow={true}
            mouseEnterDelay={0.1}
            mouseLeaveDelay={0.1}
            placement="top"
            title="白名单必定包含当前用户"
            transitionName="zoom-big-fast"
          >
            <Icon
              type="question-circle"
            />
          </Tooltip>
        </label>
        <div>
          <Switch
            defaultChecked={false}
            onChange={[Function]}
          />
        </div>
      </div>
    </div>
  </Modal>
</div>
`;
